pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2024 The Karmada Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workloadrebalancer
+
+import (
+	"context"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appsv1alpha1 "github.com/karmada-io/karmada/pkg/apis/apps/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/util"
+)
+
+func (c *RebalancerController) deleteExpiredRebalancer(key util.QueueKey) error {
+	objKey, ok := key.(client.ObjectKey)
+	if !ok {
+		klog.Errorf("Invalid object key: %+v", key)
+		return nil
+	}
+	klog.V(4).Infof("Checking if WorkloadRebalancer(%s) is ready for cleanup", objKey.Name)
+
+	// 1. Get WorkloadRebalancer from cache.
+	rebalancer := &appsv1alpha1.WorkloadRebalancer{}
+	if err := c.Client.Get(context.TODO(), objKey, rebalancer); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	// 2. Use the WorkloadRebalancer from cache to see if the TTL expires.
+	if expiredAt, err := c.processTTL(rebalancer); err != nil {
+		return err
+	} else if expiredAt == nil {
+		return nil
+	}
+
+	// 3. The WorkloadRebalancer's TTL is assumed to have expired, but the WorkloadRebalancer TTL might be stale.
+	// Before deleting the WorkloadRebalancer, do a final sanity check.
+	// If TTL is modified before we do this check, we cannot be sure if the TTL truly expires.
+	fresh, err := c.ClientWithOutCache.AppsV1alpha1().WorkloadRebalancers().Get(context.TODO(), objKey.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	// 4. Use the latest WorkloadRebalancer directly from api server to see if the TTL truly expires.
+	if expiredAt, err := c.processTTL(fresh); err != nil {
+		return err
+	} else if expiredAt == nil {
+		return nil
+	}
+
+	// 5. deletes the WorkloadRebalancer if TTL truly expires.
+	options := metav1.DeleteOptions{Preconditions: &metav1.Preconditions{ResourceVersion: &fresh.ResourceVersion}}
```

Comment:
The delete option(`ResourceVersion`) reminds me that we might not need to fetch the object from the APIServer.
Since in case of the object is rotten, the delete option would prevent this object from deleting.

/hold
Please help to run a test for it to prove it is necessary.

Reply:
done, updated